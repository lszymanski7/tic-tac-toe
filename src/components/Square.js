import React from 'react'
import PropTypes from 'prop-types'
import oIcon from '../assets/images/shapes/o.svg'
import xIcon from '../assets/images/shapes/x.svg'

const Square = ({ className, handleOnClick, index, symbol }) => {
    const icon = symbol === 'O' ? oIcon : symbol === 'X' ? xIcon : symbol

    return (
        <div className={className}>
            <button
                aria-label={`Square ${index}`}
                className="square__button"
                onClick={handleOnClick}
            >
                {icon && (
                    <img
                        alt={`${symbol} Icon`}
                        className="square__symbol"
                        src={icon}
                    />
                )}
            </button>
        </div>
    )
}

Square.propTypes = {
    className: PropTypes.string.isRequired,
    handleOnClick: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    symbol: PropTypes.string
}

export { Square as default }
