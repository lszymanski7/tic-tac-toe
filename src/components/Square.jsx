import React from 'react'
import PropTypes from 'prop-types'
import { SHAPES } from '../data/constants'
import useTheme from '../hooks/useTheme'

const Square = ({ animation, className, handleOnClick, index, symbol, winningLine }) => {
    /* istanbul ignore next */

    // Context value for dark and light theme
    const { theme } = useTheme()

    /* istanbul ignore next */

    // Function âžž sets the animation of a win, draw or game in progress
    const setAnimation = () => {
        if (typeof animation === 'object') {
            return winningLine ? animation.line : animation.notLine
        } else {
            return animation
        }
    }

    return (
        <div className={`square square--${theme}` + (className !== null ? ' ' + className : '')}>
            <button
                aria-label={`Square ${index}`}
                className="square__button"
                onClick={handleOnClick}
            >
                {symbol !== null && (
                    <img
                        alt={symbol}
                        className={`square__symbol square__symbol--${theme}` + ' ' + setAnimation()}
                        src={symbol === 'X' ? SHAPES.x : SHAPES.o}
                    />
                )}
            </button>
        </div>
    )
}

Square.propTypes = {
    animation: PropTypes.oneOfType([PropTypes.object, PropTypes.string]).isRequired,
    className: PropTypes.string,
    handleOnClick: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    symbol: PropTypes.string,
    winningLine: PropTypes.bool.isRequired
}

export default Square
