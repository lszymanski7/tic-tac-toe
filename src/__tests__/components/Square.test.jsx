import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Square from '../../components/Square'
import { ThemeProvider } from '../../contexts/ThemeContext'

const props = {
    animation: 'scale',
    className: null,
    handleOnClick: jest.fn(),
    index: 0,
    symbol: null,
    winningLine: false
}

describe('Square', () => {
    it('should be rendered correctly.', () => {
        const { container } = render(<Square {...props} />, { wrapper: ThemeProvider })
        expect(container).toMatchSnapshot()
    })

    it('should render with X symbol.', async () => {
        render(<Square {...props} symbol='X' />, { wrapper: ThemeProvider })
        const img = screen.getByRole('img')
        expect(img).toHaveAttribute('alt', 'X')
        expect(img).toHaveAttribute('src', 'x.svg')
    })

    it('should render with O symbol.', () => {
        render(<Square {...props} symbol='O' />, { wrapper: ThemeProvider })
        const img = screen.getByRole('img')
        expect(img).toHaveAttribute('alt', 'O')
        expect(img).toHaveAttribute('src', 'o.svg')
    })

    it('should handle clicking.', async () => {
        render(<Square {...props} />, { wrapper: ThemeProvider })
        const button = screen.getByRole('button')
        await userEvent.click(button)
        expect(props.handleOnClick).toHaveBeenCalled()
    })
})
