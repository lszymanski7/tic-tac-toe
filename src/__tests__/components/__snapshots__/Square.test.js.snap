// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square should be rendered correctly. 1`] = `
<div>
  <div
    class="square square--light"
  >
    <button
      aria-label="Square 0"
      class="square__button"
    />
  </div>
</div>
`;
